### PR TITLE
feat(step-6-2): improve gendiff work with nested diff

### PR DESCRIPTION
**Pull request by AlexMomot-717**

### CODE REVIEW

File: gendiff/core.py [hunk 1]
```diff
@@ -1,16 +1,14 @@
...
 
 def generate_diff(
     first_file: str, second_file: str, output_format: str = "stylish"
-) -> Any:
+) -> str:
     f1 = parse_file(first_file)
     f2 = parse_file(second_file)
     if output_format == "stylish":
-        return get_diff_view(f1, f2)
+        return "{\n" + get_diff_str_nested(f1, f2) + "}"
```
Comment (svntmr) [inline gendiff/core.py]: a function that generates diff should create `"{\n"` and `"}"` on its own - so you'll just call it to get diff string

File: gendiff/string_templates.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
+from typing import Any, Dict
```
Comment (svntmr) [inline gendiff/string_templates.py]: i'd rather see it in `formatting` module

File: gendiff/string_templates.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+
+from gendiff.diff import get_diff_view
+
+diffs_templates = {
+    "missing": "{indent}- {key}:{symb}{value}\n",
+    "added": "{indent}+ {key}:{symb}{value}\n",
+    "changed": "{indent}- {key}:{symb}{value}\n{indent}+ {key}: {ch_value}\n",
```
Comment (svntmr) [inline gendiff/string_templates.py]: what is the purpose of `symb`?

File: gendiff/string_templates.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+        )
+    else:
+        diff_str = diffs_templates["changed"].format(
+            indent=indent * " ", key=key, symb=symb, value=value, ch_value=ch_value
+        )
+
+    return diff_str
+
+
+def get_diff_str_nested(
```
Comment (svntmr) [inline gendiff/string_templates.py]: you should have three functions:  
- one that is responsible for the creation of diff (takes care of `{}`)
- one that is responsible for the processing of nested diff types
- one that is responsible for the processing of plain diff types

File: gendiff/string_templates.py [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+def flat_diff(
+    key_type: str,
+    indent: int,
+    key: str,
+    spec_char: str = " ",
+    symb: str = " ",
+    value: Any = None,
+    ch_value: Any = None,
+) -> str:
+    if key_type in ("missing", "added", "no_change"):
```
Comment (svntmr) [inline gendiff/string_templates.py]: this function should work with non-nested diff types only

**Review by svntmr — CHANGES_REQUESTED** [review]

File: gendiff/string_templates.py (hunk 3 above)
Comment (AlexMomot-717) [reply gendiff/string_templates.py]: the reason is for having whitespace after "key:" before following value (if value) or having "\n" (if not value, for instance, value = "")

File: gendiff/diff_nested.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
+from typing import Any, Dict
+
+from gendiff.diff import get_diff_view
+from gendiff.formatting import convert_value, nested_diff, plain_diff
+
+
+def get_diff_str_nested(
```
Comment (svntmr) [inline gendiff/diff_nested.py]: should be in formatting :)

File: gendiff/diff_nested.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+                d1 = dict1[key]
+                d2 = dict2[key]
+
+            elif val["type"] == "missing_nested":
+                key_type = "missing_nested"
+                d1 = d2 = dict1[key]
+            elif val["type"] == "added_nested":
+                key_type = "added_nested"
+                d1 = d2 = dict2[key]
+            else:  # "changed_nested":
```
Comment (svntmr) [inline gendiff/diff_nested.py]: please, remove `changed_nested` (I can see it underneath)

**Review by svntmr — CHANGES_REQUESTED** [review]

File: gendiff/diff_nested.py [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+    diff_str = "{\n" if indent == 2 else ""
+
+    for key, val in diff_view.items():
+        value = convert_value(val["value"])
+        ch_value = convert_value(val["changed_value"])
+        whitespace_or_not = "" if not value else " "
+
+        if val["type"].endswith("nested"):
+            if val["type"] == "nested":
+                key_type = "nested"
```
Comment (svntmr) [inline gendiff/diff_nested.py]: [optional] key_type is the same as `val["type']` (most of the time), consider using it instead

File: gendiff/formatting.py [hunk 9]
```diff
@@ -35,3 +35,58 @@ def get_diff_str(dict1: Dict[str, Any], dict2: Dict[str, Any]) -> str:
...
+    "missing_nested": "{indent}- {key}: {{\n{value}{indent}  }}\n",
+    "added_nested": "{indent}+ {key}: {{\n{value}{indent}  }}\n",
+}
+
+
+def convert_value(val: Any) -> Any:
+    if not isinstance(val, bool | None):
+        return val
+    exceptions_to_convert = {False: "false", True: "true", None: "null"}
+    return exceptions_to_convert[val]
```
Comment (svntmr) [inline gendiff/formatting.py]: can be refactored to just `return exceptions_to_convert.get(val, val)`

File: gendiff/diff_nested.py [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+            if val["type"] == "changed_nested" and isinstance(dict1[key], dict):
+                value = convert_value(dict2[key])
+                whitespace_or_not = "" if not value else " "
+                diff_str += plain_diff(
+                    key_type="added",
+                    key=key,
+                    indent=indent,
+                    symb=whitespace_or_not,
+                    value=value,
+                )
```
Comment (svntmr) [inline gendiff/diff_nested.py]: [optional] would be cool to see this part in `nested_diff`

**Review by svntmr — CHANGES_REQUESTED** [review]

File: gendiff/formatting.py [hunk 11]
```diff
@@ -35,3 +35,57 @@ def get_diff_str(dict1: Dict[str, Any], dict2: Dict[str, Any]) -> str:
...
+        plain_post = plain_diff(
+            key_type="added",
+            key=key,
+            indent=indent,
+            value=ch_value,
+        )
+        key_type = "missing_nested"
+
+    if key_type == "changed_nested_2":
+        plain_ante = plain_diff(
```
Comment (svntmr) [inline gendiff/formatting.py]: `ante`? what does it mean?

Comment (AlexMomot-717) [reply gendiff/formatting.py]: means before)))

Comment (svntmr) [reply gendiff/formatting.py]: `plain_before` & `plain_after` will be easier to read

**Review by svntmr — APPROVED** [review]

**Pull request merged**